docs/enhanced-provider-k8s.md

```diff
@@ -0,0 +1,303 @@
+
+![](images/defining-dependencies-2.png)
+
+When a managed resource and its external resource are deleted, the provider will remove the finalizers from its dependent external resources.
+
+In our case, after all `Object` resources are deleted, the `Kong` resource will be deleted at first, so that the `Kong` pod will be deleted.
+
+![](images/deleting-in-order-1.png)
+
+During the time when the `Object` resource for `Kong` is deleted, it also removes the finalizer from its dependent external resource, the `ClusterServiceVersion` resource. This triggers the actual deletion of the `ClusterServiceVersion` resource, then the deletion of `InstallPlan`. As a result, the Kong operator will be deleted.
+
+![](images/deleting-in-order-2.png)
+
+Now that the `Object` resource for `ClusterServiceVersion` is deleted, it also removes the finalizer from its dependent external resource, the `Subscription` resource. Ultimately, all resources involved in `Networking` capability will be deleted. This leaves the cluster in a clean state after the uninstall process is completed.
+
+![](images/deleting-in-order-3.png)
+
+## Summary
+
+The enhanced Provider Kubernetes introduced some interesting features that can help to manage arbitrary Kubernetes resources more effectively. It also brings more values to the provider and broarden the scope where the provider can be applied.
+
+With the help of Provider Kubernetes, different backends such as OLM, Helm, and so on that provision the actual workloads in Kubernetes cluster can be integrated into the Crossplane world easily as long as these backends work in a Kubernetes native way, i.e. driven by Kubernetes Customer Resource.
+
+Thus, we can drive the backends by creating `Object` resources that represent the corresponding customer resources watched by these backends, manage them using fine-grained policy, define references to resolve metadata at runtime, and ensure the install/uninstall order centrally and consistently. 
+
+![](images/provider-kubernetes.png)
+
+One more additional value by using enhanced Provider Kubernetes is that some features such as resource references that have not been implemeted in Crossplane core yet can be added at provider level. It tells us that if you want to try a feature but cannot wait until it can be accepted and added in Crossplane core in its official release, or even worse that may never get chance to go into Crossplane core, it would be much better to implement it in provider rather than fork the Crossplane core and modify the core directly. This is because it may be hard to track and upgrade to the following Crossplane releases based on a forked version if you want some features that are only available in the coming releases.
```

Comment:
Due to the concerns I mentioned [here](https://github.com/crossplane-contrib/provider-kubernetes/pull/14/files#r741645235), I don't think it is a good idea to give such power to provider-kubernetes. But I am wondering if it makes sense to create a separate provide just for this, e.g. `provider-reference-resolver`

Reply:
From provider user's perspective, to keep it simple and avoid confusion, ideally, there can be only single provider in community to manage Kubernetes resources. That is `provider-kubernetes`. We could turn off some features by default and only turn on them by user explicitly. To create a separate provider for reference resolving specifically and named as `provider-reference-resolver` may limit features that we would like to add in future.

Interestingly, for reference resolving, there is a valid case that I explored recently in [Using Crossplane in GitOps](https://github.com/morningspace/capabilities-shim-gitops/blob/main/docs/using-crossplane-in-gitops-3.md#dependency-resolving-by-crossplane), where user may request a neutral way to declare dependencies to ensure resource provisioning order, rather than rely on similar feature provided by specific GitOps tool, e.g. syncwave in Argo CD, so that they will not be tightly coupled with a certain GitOps tool, e.g.: as I know Flux CD does not have such thing that is equivalent to syncwave in Argo CD.

Does it make sense?

Reply:
Agreed 👍